In [1]:
import pandas as pd
import json
import numpy as np

LYRIC_FOLDER_PATH = './data/lyric-fetching-datasets'
IS_INSTRUMENTAL_RETURN_FLAG = 'THIS_IS_AN_INSTRUMENTAL'

# 1. Load the latest data

In [2]:
lyrics_df = pd.read_csv(f'{LYRIC_FOLDER_PATH}/{"2.3 OUTPUT of lyrics by song id with urls"}.csv')
lyrics_df

song_id  \
0       (I Can't Get No) SatisfactionThe Rolling Stones   
1                (I Just) Died In Your ArmsCutting Crew   
2                (I Know) I'm Losing YouThe Temptations   
3     (I've Had) The Time Of My LifeBill Medley & Je...   
4                                         1 ThingAmerie   
...                                                 ...   
5002  Saving Forever For You (From "Beverly Hills, 9...   
5003   Shakedown (From "Beverly Hills Cop II")Bob Seger   
5004  Superstar/Bless The Beasts And ChildrenCarpenters   
5005  Travelin' Band/Who'll Stop The RainCreedence C...   
5006               pride.is.the.devilJ. Cole & Lil Baby   

                                                 lyrics  \
0     (I Can’t Get No) Satisfaction Lyrics[Instrumen...   
1     (I Just) Died in Your Arms Lyrics[Intro]\nOh I...   
2     (I Know) I’m Losing You Lyrics[Intro]\nOoh\nYo...   
3     (I’ve Had) The Time of My Life Lyrics[Chorus: ...   
4     1 Thing Lyrics[Intro]\nWoo!\nUh\nWoo!\nNa-na-n...   
...                                                 ...   
5002  Saving Forever For You Lyrics[Verse 1]\nI've n...   
5003  Shakedown Lyrics[Verse 1]\nNo matter what you ...   
5004  Superstar Lyrics[Verse 1]\nLong ago and oh so ...   
5005  Travelin’ Band Lyrics[Verse 1]\n737 coming out...   
5006  TranslationsРусскийPortuguêsItaliano​​p r i d ...   

                                                   urls  
0     ["https://genius.com/The-rolling-stones-i-cant...  
1     ["https://genius.com/Cutting-crew-i-just-died-...  
2     ["https://genius.com/The-temptations-i-know-im...  
3     ["https://genius.com/Bill-medley-and-jennifer-...  
4     ["https://genius.com/Amerie-1-thing-lyrics", "...  
...                                                 ...  
5002  ["https://genius.com/Shanice-saving-forever-fo...  
5003  ["https://genius.com/Bob-seger-shakedown-lyrics"]  
5004  ["https://genius.com/Carpenters-superstar-lyri...  
5005  ["https://genius.com/Creedence-clearwater-revi...  
5006  ["https://genius.com/J-cole-and-lil-baby-p-r-i...  

[5007 rows x 3 columns]

# 2. Drop any new instrumentals found

In [3]:
NEW_INSTRUMENTALS_FOUND = [
    'Our Winter LoveBill Pursell'
]

lyrics_sans_instrumentals = lyrics_df[
    lyrics_df.song_id.apply(lambda song_id: song_id not in NEW_INSTRUMENTALS_FOUND)
]
lyrics_sans_instrumentals

song_id  \
0       (I Can't Get No) SatisfactionThe Rolling Stones   
1                (I Just) Died In Your ArmsCutting Crew   
2                (I Know) I'm Losing YouThe Temptations   
3     (I've Had) The Time Of My LifeBill Medley & Je...   
4                                         1 ThingAmerie   
...                                                 ...   
5002  Saving Forever For You (From "Beverly Hills, 9...   
5003   Shakedown (From "Beverly Hills Cop II")Bob Seger   
5004  Superstar/Bless The Beasts And ChildrenCarpenters   
5005  Travelin' Band/Who'll Stop The RainCreedence C...   
5006               pride.is.the.devilJ. Cole & Lil Baby   

                                                 lyrics  \
0     (I Can’t Get No) Satisfaction Lyrics[Instrumen...   
1     (I Just) Died in Your Arms Lyrics[Intro]\nOh I...   
2     (I Know) I’m Losing You Lyrics[Intro]\nOoh\nYo...   
3     (I’ve Had) The Time of My Life Lyrics[Chorus: ...   
4     1 Thing Lyrics[Intro]\nWoo!\nUh\nWoo!\nNa-na-n...   
...                                                 ...   
5002  Saving Forever For You Lyrics[Verse 1]\nI've n...   
5003  Shakedown Lyrics[Verse 1]\nNo matter what you ...   
5004  Superstar Lyrics[Verse 1]\nLong ago and oh so ...   
5005  Travelin’ Band Lyrics[Verse 1]\n737 coming out...   
5006  TranslationsРусскийPortuguêsItaliano​​p r i d ...   

                                                   urls  
0     ["https://genius.com/The-rolling-stones-i-cant...  
1     ["https://genius.com/Cutting-crew-i-just-died-...  
2     ["https://genius.com/The-temptations-i-know-im...  
3     ["https://genius.com/Bill-medley-and-jennifer-...  
4     ["https://genius.com/Amerie-1-thing-lyrics", "...  
...                                                 ...  
5002  ["https://genius.com/Shanice-saving-forever-fo...  
5003  ["https://genius.com/Bob-seger-shakedown-lyrics"]  
5004  ["https://genius.com/Carpenters-superstar-lyri...  
5005  ["https://genius.com/Creedence-clearwater-revi...  
5006  ["https://genius.com/J-cole-and-lil-baby-p-r-i...  

[5006 rows x 3 columns]

# 3. Apply manually fixed songs

In [4]:
fixes_raw = pd.read_csv(f'{LYRIC_FOLDER_PATH}/{"lyric_corrections_by_hand_on_v0.7_nomic_embeddings"}.csv')
fixes = fixes_raw.copy()

fixes['song_id'] = fixes.song + fixes.performer

fixes['correct_url'] = fixes.correct_genius_url
fixes = fixes[['song_id', 'correct_url']]
fixes

song_id  \
0   PeachesJustin Bieber Featuring Daniel Caesar &...   
1                     In The MoodErnie Fields & Orch.   
2        Willie And The Hand JiveThe Johnny Otis Show   
3   Sittin' Up In My Room (From "Waiting To Exhale...   
4   Streets Of Philadelphia (From "Philadelphia")B...   
5           Don't CryLil Wayne Featuring XXXTENTACION   
6              Lucky YouEminem Featuring Joyner Lucas   
7                     Wild OnesFlo Rida Featuring Sia   
8   LoyalChris Brown Featuring Lil Wayne & French ...   
9     ZEZEKodak Black Featuring Travis Scott & Offset   
10             Heat Wave/love Is A RoseLinda Ronstadt   
11            Always On TimeJa Rule Featuring Ashanti   
12     Who's Johnny ("Short Circuit" Theme)El DeBarge   
13  Baby I Love Your Way (From "Reality Bites")Big...   
14                              Mr. CusterLarry Verne   
15  FranchiseTravis Scott Featuring Young Thug & M...   
16                        Lady GodivaPeter And Gordon   
17                                   MickeyToni Basil   
18        LightersBad Meets Evil Featuring Bruno Mars   
19  Somebody That I Used To KnowGotye Featuring Ki...   
20       Maria MariaSantana Featuring The Product G&B   
21  Take What You WantPost Malone Featuring Ozzy O...   

                                          correct_url  
0     https://genius.com/Justin-bieber-peaches-lyrics  
1   https://genius.com/Ernie-fields-in-the-mood-ly...  
2   https://genius.com/Johnny-otis-willie-and-the-...  
3   https://genius.com/Brandy-sittin-up-in-my-room...  
4        https://genius.com/Lil-wayne-dont-cry-lyrics  
5        https://genius.com/Lil-wayne-dont-cry-lyrics  
6          https://genius.com/Eminem-lucky-you-lyrics  
7        https://genius.com/Flo-rida-wild-ones-lyrics  
8         https://genius.com/Chris-brown-loyal-lyrics  
9          https://genius.com/Kodak-black-zeze-lyrics  
10  https://genius.com/Linda-ronstadt-heat-wave-ly...  
11   https://genius.com/Ja-rule-always-on-time-lyrics  
12   https://genius.com/El-debarge-whos-johnny-lyrics  
13  https://genius.com/Big-mountain-baby-i-love-yo...  
14  https://genius.com/Larry-verne-mister-custer-l...  
15  https://genius.com/Travis-scott-franchise-remi...  
16  https://genius.com/Peter-and-gordon-lady-godiv...  
17        https://genius.com/Toni-basil-mickey-lyrics  
18  https://genius.com/Bad-meets-evil-lighters-lyrics  
19  https://genius.com/Gotye-somebody-that-i-used-...  
20      https://genius.com/Santana-maria-maria-lyrics  
21  https://genius.com/Post-malone-take-what-you-w...

In [5]:
import lyricsgenius
GENIUS_CLIEN_API_KEY = '8ZiCNVQqbSF6NM0nwVnVv6Y0EBR7BsVQ9HaVgPdqJOZR1l6XlymR-pUUi6z3GHPx' # be nice :)
genius = lyricsgenius.Genius(GENIUS_CLIEN_API_KEY)

In [6]:
def _get_song(url):
    try: 
        print(url, '...')
        return [genius.lyrics(song_url=url), [url]]
    except:
        print(f'OOPS: {url}')
        
def refetch_lyrics_from_url(songs_subset_df):
    return songs_subset_df.apply(
        lambda row: _get_song(row['correct_url']),
        axis=1
    )

# fetched = refetch_lyrics_from_url(fixes)
# fetched

In [7]:
def convert_fetched_lyrics_to_df(fetched_lyrics_and_url_options_arrays, original_df):
    just_lyrics = [x and x[0] for x in fetched_lyrics_and_url_options_arrays]
    just_stringified_urls = [x and json.dumps(x[1]) for x in fetched_lyrics_and_url_options_arrays]
    save_to_disk = pd.DataFrame({
        'song_id': original_df.song_id, 'lyrics': just_lyrics, 'urls': just_stringified_urls
    })
    return save_to_disk

# fixes_fetched_df = convert_fetched_lyrics_to_df(fetched, fixes)
# fixes_fetched_df

In [8]:
def save_lyrics(lyrics_subset, file_name):
    lyrics_subset.to_csv(f'./data/lyric-fetching-datasets/{file_name}.csv', index=False)
    return lyrics_subset

In [9]:
FIXES_TITLE = 'FETCHED LYRICS from lyric_corrections_by_hand_on_v0.7_nomic_embeddings'
# save_lyrics(fixes_fetched_df, FIXES_TITLE)

fixes_fetched_df = pd.read_csv(f'{LYRIC_FOLDER_PATH}/{FIXES_TITLE}.csv')
fixes_fetched_df

song_id  \
0   PeachesJustin Bieber Featuring Daniel Caesar &...   
1                     In The MoodErnie Fields & Orch.   
2        Willie And The Hand JiveThe Johnny Otis Show   
3   Sittin' Up In My Room (From "Waiting To Exhale...   
4   Streets Of Philadelphia (From "Philadelphia")B...   
5           Don't CryLil Wayne Featuring XXXTENTACION   
6              Lucky YouEminem Featuring Joyner Lucas   
7                     Wild OnesFlo Rida Featuring Sia   
8   LoyalChris Brown Featuring Lil Wayne & French ...   
9     ZEZEKodak Black Featuring Travis Scott & Offset   
10             Heat Wave/love Is A RoseLinda Ronstadt   
11            Always On TimeJa Rule Featuring Ashanti   
12     Who's Johnny ("Short Circuit" Theme)El DeBarge   
13  Baby I Love Your Way (From "Reality Bites")Big...   
14                              Mr. CusterLarry Verne   
15  FranchiseTravis Scott Featuring Young Thug & M...   
16                        Lady GodivaPeter And Gordon   
17                                   MickeyToni Basil   
18        LightersBad Meets Evil Featuring Bruno Mars   
19  Somebody That I Used To KnowGotye Featuring Ki...   
20       Maria MariaSantana Featuring The Product G&B   
21  Take What You WantPost Malone Featuring Ozzy O...   

                                               lyrics  \
0   TranslationsTürkçeEspañolPortuguês日本語ItalianoD...   
1   In The Mood LyricsWho's the livin' dolly with ...   
2   Willie and the Hand Jive LyricsI know a cat na...   
3   Sittin’ Up In My Room Lyrics[Verse 1]\nSeems l...   
4   TranslationsTürkçeРусскийFrançaisPortuguêsDon’...   
5   TranslationsTürkçeРусскийFrançaisPortuguêsDon’...   
6   TranslationsTürkçeEminem - Lucky You ft. Joyne...   
7   TranslationsPortuguêsWild Ones Lyrics[Chorus: ...   
8   Loyal Lyrics[Intro: Lil Wayne, Chris Brown]\nY...   
9   TranslationsPortuguêsZEZE Lyrics[Intro]\nD.A. ...   
10  Heat Wave LyricsWhenever I'm with him\nSomethi...   
11  Always on Time Lyrics[Intro: Ashanti]\nAlways ...   
12  Who’s Johnny LyricsThere she goes and knows I'...   
13  Baby, I Love Your Way Lyrics[Intro]\nOoh, baby...   
14  Mister Custer LyricsThat famous day in history...   
15  TranslationsEnglishFRANCHISE (REMIX) Lyrics[In...   
16  Lady Godiva Lyrics[Chorus]\nSeventeen, a beaut...   
17  Mickey Lyrics[Refrain]\nOh, Mickey\nYou're so ...   
18  Lighters Lyrics[Chorus: Bruno Mars]\nThis one'...   
19  TranslationsTürkçeSomebody That I Used to Know...   
20  Maria Maria Lyrics[Intro: Wyclef Jean]\nLadies...   
21  TranslationsSvenskaEspañolPortuguêsPolskiPost ...   

                                                 urls  
0   ["https://genius.com/Justin-bieber-peaches-lyr...  
1   ["https://genius.com/Ernie-fields-in-the-mood-...  
2   ["https://genius.com/Johnny-otis-willie-and-th...  
3   ["https://genius.com/Brandy-sittin-up-in-my-ro...  
4    ["https://genius.com/Lil-wayne-dont-cry-lyrics"]  
5    ["https://genius.com/Lil-wayne-dont-cry-lyrics"]  
6      ["https://genius.com/Eminem-lucky-you-lyrics"]  
7    ["https://genius.com/Flo-rida-wild-ones-lyrics"]  
8     ["https://genius.com/Chris-brown-loyal-lyrics"]  
9      ["https://genius.com/Kodak-black-zeze-lyrics"]  
10  ["https://genius.com/Linda-ronstadt-heat-wave-...  
11  ["https://genius.com/Ja-rule-always-on-time-ly...  
12  ["https://genius.com/El-debarge-whos-johnny-ly...  
13  ["https://genius.com/Big-mountain-baby-i-love-...  
14  ["https://genius.com/Larry-verne-mister-custer...  
15  ["https://genius.com/Travis-scott-franchise-re...  
16  ["https://genius.com/Peter-and-gordon-lady-god...  
17    ["https://genius.com/Toni-basil-mickey-lyrics"]  
18  ["https://genius.com/Bad-meets-evil-lighters-l...  
19  ["https://genius.com/Gotye-somebody-that-i-use...  
20  ["https://genius.com/Santana-maria-maria-lyrics"]  
21  ["https://genius.com/Post-malone-take-what-you...

In [10]:
JUST_SONG_ID_AND_LYRICS_COLS = ['song_id', 'lyrics']
SONG_ID = 'song_id'

def remove_overlap(base_df, overlap_df_to_remove):
    return base_df[
        base_df.song_id.apply(
            lambda i: i not in overlap_df_to_remove.song_id.values
        )
    ]

In [11]:
# sanity check
fixes_fetched_df[
    fixes_fetched_df.song_id.apply(lambda i: i not in lyrics_sans_instrumentals.song_id.values)
]

Empty DataFrame
Columns: [song_id, lyrics, urls]
Index: []

### swap back into main dataset

In [12]:
lyrics_sans_ins_with_fixes = pd.concat([
    remove_overlap(lyrics_sans_instrumentals, fixes_fetched_df),
    fixes_fetched_df
]).reset_index(drop=True)
lyrics_sans_ins_with_fixes

song_id  \
0       (I Can't Get No) SatisfactionThe Rolling Stones   
1                (I Just) Died In Your ArmsCutting Crew   
2                (I Know) I'm Losing YouThe Temptations   
3     (I've Had) The Time Of My LifeBill Medley & Je...   
4                                         1 ThingAmerie   
...                                                 ...   
5001                                   MickeyToni Basil   
5002        LightersBad Meets Evil Featuring Bruno Mars   
5003  Somebody That I Used To KnowGotye Featuring Ki...   
5004       Maria MariaSantana Featuring The Product G&B   
5005  Take What You WantPost Malone Featuring Ozzy O...   

                                                 lyrics  \
0     (I Can’t Get No) Satisfaction Lyrics[Instrumen...   
1     (I Just) Died in Your Arms Lyrics[Intro]\nOh I...   
2     (I Know) I’m Losing You Lyrics[Intro]\nOoh\nYo...   
3     (I’ve Had) The Time of My Life Lyrics[Chorus: ...   
4     1 Thing Lyrics[Intro]\nWoo!\nUh\nWoo!\nNa-na-n...   
...                                                 ...   
5001  Mickey Lyrics[Refrain]\nOh, Mickey\nYou're so ...   
5002  Lighters Lyrics[Chorus: Bruno Mars]\nThis one'...   
5003  TranslationsTürkçeSomebody That I Used to Know...   
5004  Maria Maria Lyrics[Intro: Wyclef Jean]\nLadies...   
5005  TranslationsSvenskaEspañolPortuguêsPolskiPost ...   

                                                   urls  
0     ["https://genius.com/The-rolling-stones-i-cant...  
1     ["https://genius.com/Cutting-crew-i-just-died-...  
2     ["https://genius.com/The-temptations-i-know-im...  
3     ["https://genius.com/Bill-medley-and-jennifer-...  
4     ["https://genius.com/Amerie-1-thing-lyrics", "...  
...                                                 ...  
5001    ["https://genius.com/Toni-basil-mickey-lyrics"]  
5002  ["https://genius.com/Bad-meets-evil-lighters-l...  
5003  ["https://genius.com/Gotye-somebody-that-i-use...  
5004  ["https://genius.com/Santana-maria-maria-lyrics"]  
5005  ["https://genius.com/Post-malone-take-what-you...  

[5006 rows x 3 columns]

# 4. Refetching lyrics that are likely wrong

### Here: we demonstrate how `/` songs are usually older tracks that had a physical A & B side when sold, & break when trying to be fetched from genius.com

In [13]:
lyrics_sans_ins_with_fixes[
    lyrics_sans_ins_with_fixes.song_id.str.contains('/') 
#     lyrics_sans_ins_with_fixes.song_id.str.contains("Lookin' Out My Back Door/Long As I Can See The Light")
#     lyrics_sans_ins_with_fixes.lyrics.str.contains("Greatest Songs")
    
]

song_id  \
70                             Dear Mama/Old School2Pac   
155   How Do U Want It/California Love2Pac Featuring...   
190             I Am...I Said/Done Too SoonNeil Diamond   
226   I Don't Wanna Live Forever (Fifty Shades Darke...   
245                         I Got Id/Long RoadPearl Jam   
...                                                 ...   
4975  Down On The Corner/Fortunate SonCreedence Clea...   
4978   Rainy Night In Georgia/Rubberneckin'Brook Benton   
4981  Superstar/Bless The Beasts And ChildrenCarpenters   
4982  Travelin' Band/Who'll Stop The RainCreedence C...   
4994             Heat Wave/love Is A RoseLinda Ronstadt   

                                                 lyrics  \
70    TranslationsPortuguêsDeutschEnglish2Pac - Dear...   
155   Hot Rap Songs Chart 25th Anniversary: Top 100 ...   
190   Songs of Eulogy and Bereavement LyricsListed i...   
226   New Music Friday 12/16/16 LyricsZayn & Taylor ...   
245   Even Flow - Live in Seattle, WA 06-December-20...   
...                                                 ...   
4975  Down on the Corner Lyrics[Verse 1]\nEarly in t...   
4978  Rainy Night in Georgia Lyrics[Verse 1]\nHoveri...   
4981  Superstar Lyrics[Verse 1]\nLong ago and oh so ...   
4982  Travelin’ Band Lyrics[Verse 1]\n737 coming out...   
4994  Heat Wave LyricsWhenever I'm with him\nSomethi...   

                                                   urls  
70    ["https://genius.com/2pac-dear-mama-lyrics", "...  
155   ["https://genius.com/Billboard-hot-rap-songs-c...  
190   ["https://genius.com/Genius-lists-songs-of-eul...  
226   ["https://genius.com/Spotify-new-music-friday-...  
245   ["https://genius.com/Pearl-jam-even-flow-live-...  
...                                                 ...  
4975  ["https://genius.com/Creedence-clearwater-revi...  
4978  ["https://genius.com/Brook-benton-rainy-night-...  
4981  ["https://genius.com/Carpenters-superstar-lyri...  
4982  ["https://genius.com/Creedence-clearwater-revi...  
4994  ["https://genius.com/Linda-ronstadt-heat-wave-...  

[77 rows x 3 columns]

In [14]:
import json

In [15]:
has_url = lyrics_sans_ins_with_fixes[
    lyrics_sans_ins_with_fixes.urls.notna()
].copy().reset_index(drop=True)
has_url['url'] = has_url.urls.apply(lambda url: url and json.loads(url)[0])
has_url

song_id  \
0       (I Can't Get No) SatisfactionThe Rolling Stones   
1                (I Just) Died In Your ArmsCutting Crew   
2                (I Know) I'm Losing YouThe Temptations   
3     (I've Had) The Time Of My LifeBill Medley & Je...   
4                                         1 ThingAmerie   
...                                                 ...   
4933                                   MickeyToni Basil   
4934        LightersBad Meets Evil Featuring Bruno Mars   
4935  Somebody That I Used To KnowGotye Featuring Ki...   
4936       Maria MariaSantana Featuring The Product G&B   
4937  Take What You WantPost Malone Featuring Ozzy O...   

                                                 lyrics  \
0     (I Can’t Get No) Satisfaction Lyrics[Instrumen...   
1     (I Just) Died in Your Arms Lyrics[Intro]\nOh I...   
2     (I Know) I’m Losing You Lyrics[Intro]\nOoh\nYo...   
3     (I’ve Had) The Time of My Life Lyrics[Chorus: ...   
4     1 Thing Lyrics[Intro]\nWoo!\nUh\nWoo!\nNa-na-n...   
...                                                 ...   
4933  Mickey Lyrics[Refrain]\nOh, Mickey\nYou're so ...   
4934  Lighters Lyrics[Chorus: Bruno Mars]\nThis one'...   
4935  TranslationsTürkçeSomebody That I Used to Know...   
4936  Maria Maria Lyrics[Intro: Wyclef Jean]\nLadies...   
4937  TranslationsSvenskaEspañolPortuguêsPolskiPost ...   

                                                   urls  \
0     ["https://genius.com/The-rolling-stones-i-cant...   
1     ["https://genius.com/Cutting-crew-i-just-died-...   
2     ["https://genius.com/The-temptations-i-know-im...   
3     ["https://genius.com/Bill-medley-and-jennifer-...   
4     ["https://genius.com/Amerie-1-thing-lyrics", "...   
...                                                 ...   
4933    ["https://genius.com/Toni-basil-mickey-lyrics"]   
4934  ["https://genius.com/Bad-meets-evil-lighters-l...   
4935  ["https://genius.com/Gotye-somebody-that-i-use...   
4936  ["https://genius.com/Santana-maria-maria-lyrics"]   
4937  ["https://genius.com/Post-malone-take-what-you...   

                                                    url  
0     https://genius.com/The-rolling-stones-i-cant-g...  
1     https://genius.com/Cutting-crew-i-just-died-in...  
2     https://genius.com/The-temptations-i-know-im-l...  
3     https://genius.com/Bill-medley-and-jennifer-wa...  
4              https://genius.com/Amerie-1-thing-lyrics  
...                                                 ...  
4933        https://genius.com/Toni-basil-mickey-lyrics  
4934  https://genius.com/Bad-meets-evil-lighters-lyrics  
4935  https://genius.com/Gotye-somebody-that-i-used-...  
4936      https://genius.com/Santana-maria-maria-lyrics  
4937  https://genius.com/Post-malone-take-what-you-w...  

[4938 rows x 4 columns]

In [16]:
songs_df = pd.read_csv('./data/1 DONE RIGHT OUTPUT unique songs.csv')

In [17]:
songs_df

song_id  chart_position  \
0                                   #9 DreamJohn Lennon               9   
1     '03 Bonnie & ClydeJay-Z Featuring Beyonce Knowles               4   
2                             '65 Love AffairPaul Davis               6   
3                ('til) I Kissed YouThe Everly Brothers               4   
4     (Can't Live Without Your) Love And AffectionNe...               1   
...                                                 ...             ...   
5061                   everything i wantedBillie Eilish               8   
5062                      iSpyKYLE Featuring Lil Yachty               4   
5063                                   interludeJ. Cole               8   
5064                 my.lifeJ. Cole, 21 Savage & Morray               2   
5065               pride.is.the.devilJ. Cole & Lil Baby               7   

     chart_debut                                          song  \
0     1974-12-21                                      #9 Dream   
1     2002-10-26                            '03 Bonnie & Clyde   
2     1982-02-27                               '65 Love Affair   
3     1959-08-15                           ('til) I Kissed You   
4     1990-07-07  (Can't Live Without Your) Love And Affection   
...          ...                                           ...   
5061  2019-11-23                           everything i wanted   
5062  2017-01-14                                          iSpy   
5063  2021-05-22                                     interlude   
5064  2021-05-29                                       my.life   
5065  2021-05-29                            pride.is.the.devil   

                            performer                        artist_array  \
0                         John Lennon                     ['John Lennon']   
1     Jay-Z Featuring Beyonce Knowles        ['Jay-Z', 'Beyonce Knowles']   
2                          Paul Davis                      ['Paul Davis']   
3                 The Everly Brothers             ['The Everly Brothers']   
4                              Nelson                          ['Nelson']   
...                               ...                                 ...   
5061                    Billie Eilish                   ['Billie Eilish']   
5062        KYLE Featuring Lil Yachty              ['KYLE', 'Lil Yachty']   
5063                          J. Cole                         ['J. Cole']   
5064      J. Cole, 21 Savage & Morray  ['J. Cole', '21 Savage', 'Morray']   
5065               J. Cole & Lil Baby             ['J. Cole', 'Lil Baby']   

     generic_genre  
0             ROCK  
1          HIP_HOP  
2             ROCK  
3              POP  
4          COUNTRY  
...            ...  
5061           POP  
5062           POP  
5063       HIP_HOP  
5064       HIP_HOP  
5065       HIP_HOP  

[5066 rows x 7 columns]

### Here, find songs that are `likely_wrong` simply because they don't end with `-lyrics`

In [18]:
likely_wrong = has_url[
    ~has_url.url.str.endswith('-lyrics')
]
likely_wrong = likely_wrong.merge(on='song_id', how='left', right=songs_df)[['song_id', 'song', 'performer']]
likely_wrong

song_id  \
0        Hungry Eyes (From "Dirty Dancing")Eric Carmen   
1    HungryPaul Revere & The Raiders Featuring Mark...   
2              I Am...I Said/Done Too SoonNeil Diamond   
3    I Believe In You And Me (From "The Preacher's ...   
4            I Can't Stand ItEric Clapton And His Band   
..                                                 ...   
174  In The Still Of The Nite (From "The Jacksons")...   
175      Junior's Farm/Sally GPaul McCartney And Wings   
176  Keep On, Keepin' On (From "Sunset Park")MC Lyt...   
177          O-o-h Child/Dear PrudenceThe 5 Stairsteps   
178     On Our Own (From "Ghostbusters II")Bobby Brown   

                                                  song  \
0                   Hungry Eyes (From "Dirty Dancing")   
1                                               Hungry   
2                          I Am...I Said/Done Too Soon   
3    I Believe In You And Me (From "The Preacher's ...   
4                                     I Can't Stand It   
..                                                 ...   
174     In The Still Of The Nite (From "The Jacksons")   
175                              Junior's Farm/Sally G   
176           Keep On, Keepin' On (From "Sunset Park")   
177                          O-o-h Child/Dear Prudence   
178                On Our Own (From "Ghostbusters II")   

                                            performer  
0                                         Eric Carmen  
1    Paul Revere & The Raiders Featuring Mark Lindsay  
2                                        Neil Diamond  
3                                     Whitney Houston  
4                           Eric Clapton And His Band  
..                                                ...  
174                                       Boyz II Men  
175                          Paul McCartney And Wings  
176                          MC Lyte Featuring Xscape  
177                                  The 5 Stairsteps  
178                                       Bobby Brown  

[179 rows x 3 columns]

In [19]:
BANNED_CHARACTERS = ['[', ']', '"', "'", '+', ',', '.', '(', ')', '/', '\\', '-', '&', '?', '!']
def clean_it(s):
    cleaned = s
    for naughty_char in BANNED_CHARACTERS:
        cleaned = cleaned.replace(naughty_char, '')
    return cleaned.strip()

def clean_artists(artist_str):
    s = artist_str.split('(')[0]
    s = s.split(' and ')[0]
    s = s.split(' with ')[0]
    s = s.split(' With ')[0]
    s = s.split(' And ')[0]
    s = s.split(' Featuring ')[0]
    return clean_it(s)

def get_search_term(_song, _performer):
    song = clean_it(_song.split('(')[0].split('/')[0])
    performer = clean_artists(_performer)
    return f'{performer} {song}'

likely_wrong.apply(lambda row: get_search_term(row['song'], row['performer']), axis=1).values

array(['Eric Carmen Hungry Eyes', 'Paul Revere  The Raiders Hungry',
       'Neil Diamond I AmI Said',
       'Whitney Houston I Believe In You And Me',
       'Eric Clapton I Cant Stand It',
       'Archie Bell  The Drells I Cant Stop Dancing',
       'Paul Anka I Dont Like To Sleep Alone',
       'Zayn  Taylor Swift I Dont Wanna Live Forever',
       'The Black Eyed Peas I Gotta Feeling', 'P Diddy I Need A Girl',
       'The Partridge Family Starring Shirley Jones I Think I Love You',
       'Paul Anka ', 'Logic 18002738255',
       'Ariana Grande Feat Doja Cat  Megan Thee Stallion 3435',
       'Cher  Peter Cetera After All', 'Don McLean American Pie',
       'The Guess Who American Woman', 'Paul McCartney Another Day',
       'Adam Wade As If I Didnt Know',
       'Justin Bieber As Long As You Love Me', 'Inner Circle Bad Boys',
       'Machine Gun Kelly x Camila Cabello Bad Things',
       'Monica Before You Walk Out Of My Life', 'Chicago Beginnings',
       'Jon Bon Jovi Blaze Of 

## Refetch likely wrong songs, but more smartly

In [20]:
def search_song_chose_first_then_get_lyrics(song, performer):
    try: 
        print(get_search_term(song, performer))
        songs = genius.search_songs(get_search_term(song, performer))
        urls = [song['result']['url'] for song in songs['hits']]
        print(urls)
        return [genius.lyrics(song_url=urls[0]), urls]
    except:
        print(f'OOPS: {performer} {song}')
        
        
def fetch_lyrics_and_urls(songs_subset_df):
    return songs_subset_df.apply(
        lambda row: search_song_chose_first_then_get_lyrics(song=row['song'], performer=row['performer']),
        axis=1
    )

In [21]:
# refetch_res = fetch_lyrics_and_urls(likely_wrong)

In [22]:
# _refetched_df = convert_fetched_lyrics_to_df(refetch_res, likely_wrong)
_TITLE = 'initial refetch of 179 from full that did not end in -lyrics'
# save_lyrics(_refetched_df, _TITLE)

_refetched_df = pd.read_csv(f'{LYRIC_FOLDER_PATH}/{_TITLE}.csv')
def convert_to_url(urls):
    try:
        return json.loads(urls)
    except:
        return None
_refetched_df['urls'] = _refetched_df.urls.apply(lambda urls: convert_to_url(urls))
_refetched_df

song_id  \
0        Hungry Eyes (From "Dirty Dancing")Eric Carmen   
1    HungryPaul Revere & The Raiders Featuring Mark...   
2              I Am...I Said/Done Too SoonNeil Diamond   
3    I Believe In You And Me (From "The Preacher's ...   
4            I Can't Stand ItEric Clapton And His Band   
..                                                 ...   
174  In The Still Of The Nite (From "The Jacksons")...   
175      Junior's Farm/Sally GPaul McCartney And Wings   
176  Keep On, Keepin' On (From "Sunset Park")MC Lyt...   
177          O-o-h Child/Dear PrudenceThe 5 Stairsteps   
178     On Our Own (From "Ghostbusters II")Bobby Brown   

                                                lyrics  \
0    Hungry Eyes Lyrics[Verse 1]\nI've been meaning...   
1    Hungry Lyrics[Verse 1]\nGirl, you got this nee...   
2    I Am... I Said Lyrics[Verse 1]\nL.A.'s fine, t...   
3    I Believe In You and Me Lyrics[Verse 1]\nI bel...   
4    I Can’t Stand It Lyrics[Verse 1]\nYou've been ...   
..                                                 ...   
174  In the Still of the Nite (I’ll Remember) Lyric...   
175  Junior’s Farm (DJ edit) LyricsAh, you should h...   
176  Keep On Keepin’ On Lyrics[Verse 1: MC Lyte]\nB...   
177  The 500 Greatest Songs of All Time Lyrics1. Li...   
178  On Our Own Lyrics[Intro]\nYeah..\nYou want som...   

                                                  urls  
0    [https://genius.com/Eric-carmen-hungry-eyes-ly...  
1    [https://genius.com/Paul-revere-and-the-raider...  
2    [https://genius.com/Neil-diamond-i-am-i-said-l...  
3    [https://genius.com/Whitney-houston-i-believe-...  
4    [https://genius.com/Eric-clapton-i-cant-stand-...  
..                                                 ...  
174  [https://genius.com/Boyz-ii-men-in-the-still-o...  
175  [https://genius.com/Paul-mccartney-juniors-far...  
176  [https://genius.com/Mc-lyte-keep-on-keepin-on-...  
177  [https://genius.com/Rolling-stone-the-500-grea...  
178  [https://genius.com/Bobby-brown-on-our-own-lyr...  

[179 rows x 3 columns]

## Filter out *just* the `probably_correct_now` data: (1) fetched succeeded (2) URL ends with `-lyrics`

In [23]:
_worked = _refetched_df[
    _refetched_df.urls.notna()
]
probably_correct_now = _worked[
    _worked.urls.apply(lambda urls: urls and urls[0].endswith('-lyrics'))
]
probably_correct_now

song_id  \
0        Hungry Eyes (From "Dirty Dancing")Eric Carmen   
1    HungryPaul Revere & The Raiders Featuring Mark...   
2              I Am...I Said/Done Too SoonNeil Diamond   
3    I Believe In You And Me (From "The Preacher's ...   
4            I Can't Stand ItEric Clapton And His Band   
..                                                 ...   
173    I'd Die Without You (From "Boomerang")P.M. Dawn   
174  In The Still Of The Nite (From "The Jacksons")...   
175      Junior's Farm/Sally GPaul McCartney And Wings   
176  Keep On, Keepin' On (From "Sunset Park")MC Lyt...   
178     On Our Own (From "Ghostbusters II")Bobby Brown   

                                                lyrics  \
0    Hungry Eyes Lyrics[Verse 1]\nI've been meaning...   
1    Hungry Lyrics[Verse 1]\nGirl, you got this nee...   
2    I Am... I Said Lyrics[Verse 1]\nL.A.'s fine, t...   
3    I Believe In You and Me Lyrics[Verse 1]\nI bel...   
4    I Can’t Stand It Lyrics[Verse 1]\nYou've been ...   
..                                                 ...   
173  I’d Die Without You Lyrics[Verse 1]\nIs it my ...   
174  In the Still of the Nite (I’ll Remember) Lyric...   
175  Junior’s Farm (DJ edit) LyricsAh, you should h...   
176  Keep On Keepin’ On Lyrics[Verse 1: MC Lyte]\nB...   
178  On Our Own Lyrics[Intro]\nYeah..\nYou want som...   

                                                  urls  
0    [https://genius.com/Eric-carmen-hungry-eyes-ly...  
1    [https://genius.com/Paul-revere-and-the-raider...  
2    [https://genius.com/Neil-diamond-i-am-i-said-l...  
3    [https://genius.com/Whitney-houston-i-believe-...  
4    [https://genius.com/Eric-clapton-i-cant-stand-...  
..                                                 ...  
173  [https://genius.com/Pm-dawn-id-die-without-you...  
174  [https://genius.com/Boyz-ii-men-in-the-still-o...  
175  [https://genius.com/Paul-mccartney-juniors-far...  
176  [https://genius.com/Mc-lyte-keep-on-keepin-on-...  
178  [https://genius.com/Bobby-brown-on-our-own-lyr...  

[123 rows x 3 columns]

## ... now merge in JUST the songs that we (likely) corrected succesfully

In [24]:
ready_for_manual_fixes = pd.concat([
    remove_overlap(lyrics_sans_ins_with_fixes, likely_wrong),
    probably_correct_now
]).reset_index(drop=True)
ready_for_manual_fixes

song_id  \
0       (I Can't Get No) SatisfactionThe Rolling Stones   
1                (I Just) Died In Your ArmsCutting Crew   
2                (I Know) I'm Losing YouThe Temptations   
3     (I've Had) The Time Of My LifeBill Medley & Je...   
4                                         1 ThingAmerie   
...                                                 ...   
4945    I'd Die Without You (From "Boomerang")P.M. Dawn   
4946  In The Still Of The Nite (From "The Jacksons")...   
4947      Junior's Farm/Sally GPaul McCartney And Wings   
4948  Keep On, Keepin' On (From "Sunset Park")MC Lyt...   
4949     On Our Own (From "Ghostbusters II")Bobby Brown   

                                                 lyrics  \
0     (I Can’t Get No) Satisfaction Lyrics[Instrumen...   
1     (I Just) Died in Your Arms Lyrics[Intro]\nOh I...   
2     (I Know) I’m Losing You Lyrics[Intro]\nOoh\nYo...   
3     (I’ve Had) The Time of My Life Lyrics[Chorus: ...   
4     1 Thing Lyrics[Intro]\nWoo!\nUh\nWoo!\nNa-na-n...   
...                                                 ...   
4945  I’d Die Without You Lyrics[Verse 1]\nIs it my ...   
4946  In the Still of the Nite (I’ll Remember) Lyric...   
4947  Junior’s Farm (DJ edit) LyricsAh, you should h...   
4948  Keep On Keepin’ On Lyrics[Verse 1: MC Lyte]\nB...   
4949  On Our Own Lyrics[Intro]\nYeah..\nYou want som...   

                                                   urls  
0     ["https://genius.com/The-rolling-stones-i-cant...  
1     ["https://genius.com/Cutting-crew-i-just-died-...  
2     ["https://genius.com/The-temptations-i-know-im...  
3     ["https://genius.com/Bill-medley-and-jennifer-...  
4     ["https://genius.com/Amerie-1-thing-lyrics", "...  
...                                                 ...  
4945  [https://genius.com/Pm-dawn-id-die-without-you...  
4946  [https://genius.com/Boyz-ii-men-in-the-still-o...  
4947  [https://genius.com/Paul-mccartney-juniors-far...  
4948  [https://genius.com/Mc-lyte-keep-on-keepin-on-...  
4949  [https://genius.com/Bobby-brown-on-our-own-lyr...  

[4950 rows x 3 columns]

### We'll deal with these below:

In [25]:
ready_for_manual_fixes.urls.isna().sum()

68

## Inspect the "fishy" non `-lyric` ending urls

In [26]:
_non_lyrics_word_url_ending = remove_overlap(_worked, probably_correct_now)
_non_lyrics_word_url_ending
# [print(x) for x in _non_lyrics_word_url_ending.apply(lambda row: row['song_id'] + '\n' + row['urls'][0] + '\n', axis=1).values]
# _non_lyrics_word_url_ending.urls.apply(lambda urls: urls[0]).values

song_id  \
5         I Can't Stop DancingArchie Bell & The Drells   
8                   I Gotta FeelingThe Black Eyed Peas   
10   I Think I Love YouThe Partridge Family Starrin...   
13   34+35Ariana Grande Feat. Doja Cat & Megan Thee...   
18                        As If I Didn't KnowAdam Wade   
21        Bad ThingsMachine Gun Kelly x Camila Cabello   
26                    Boom Boom PowThe Black Eyed Peas   
33                         Convention '72The Delegates   
39         Do We Have A Problem?Nicki Minaj X Lil Baby   
40   Doesn't Somebody Want To Be WantedThe Partridg...   
46                          EnemyImagine Dragons X JID   
48                     Every 1's A WinnerHot Chocolate   
49               Everyday PeopleSly & The Family Stone   
58   Gonna Make You Sweat (Everybody Dance Now)C+C ...   
64             Here We GoC+C Music Factory/F. Williams   
70              I'll Never Fall In Love AgainTom Jones   
73                 ImagineJohn Lennon Plastic Ono Band   
75                      It Ain't MeKygo x Selena Gomez   
82                            LaylaDerek & The Dominos   
83          Lean OnMajor Lazer & DJ Snake Featuring MO   
89             Lost In YouGarth Brooks as Chris Gaines   
92   Love Theme From Romeo & JulietHenry Mancini An...   
93                  Love Won't Let Me WaitMajor Harris   
96                   Me, Myself & IG-Eazy x Bebe Rexha   
101       More, More, More Pt. 1Andrea True Connection   
102                        MoreKai Winding & Orchestra   
112            Philadelphia FreedomThe Elton John Band   
115     Quiet VillageThe Exotic Sounds of Martin Denny   
117          Release MeEsther Phillips "Little Esther"   
122                              Sadeness Part 1Enigma   
125  Say SomethingA Great Big World & Christina Agu...   
128  Shame On The MoonBob Seger & The Silver Bullet...   
130                        Simon Says1910 Fruitgum Co.   
134   Still The SameBob Seger & The Silver Bullet Band   
135                Stranger On The ShoreMr. Acker Bilk   
137             Supernatural Thing - Part IBen E. King   
141    The Ballad Of The Green BeretsSSgt Barry Sadler   
142                       The Big HurtMiss Toni Fisher   
147  The Theme From Hill Street BluesMike Post feat...   
149  Theme From "Greatest American Hero" (Believe I...   
159              Where Is The Love?The Black Eyed Peas   
162             Will You Love Me TomorrowThe Shirelles   
165                      Yellow BirdArthur Lyman Group   
177          O-o-h Child/Dear PrudenceThe 5 Stairsteps   

                                                lyrics  \
5    I Can’t Stop Dancing Lyrics(Oh, no, I just can...   
8    Soundtrack Of My Life Lyrics1. I am born (1998...   
10   June 2021 Singles Release Calendar Lyrics6/1\n...   
13   Doja Cat [Discography List] LyricsStudio Album...   
18   Lady Chatterley’s Lover (Chap. 16) LyricsConni...   
21   This Is Machine Gun Kelly LyricsUpdate Playlis...   
26   Nostalgic Songs LyricsElementary School and Yo...   
33   State of the Union 1858 Lyrics                ...   
39   Today’s Top Hits 2/4/22 (ft. Lil Nas X) Lyrics...   
40   Angela’s Ashes LyricsAngela's Ashes\n\nA Memoi...   
46   Today’s Top Hits 2/4/22 (ft. Lil Nas X) Lyrics...   
48   Scarface Script LyricsScarface\n\nBy: Oliver S...   
49   Black History Salute Lyrics* = Page created bu...   
58   2021: New Year, Same Sad LyricsYOUTUBE PLAYLIS...   
64   Letter to the Administration LyricsDear Grambl...   
70   I’ll Never Fall in Love Again Lyrics[Vers 1]\n...   
73   ​vwm’s immaculate collection LyricsVinylA Cert...   
75   ​youth group - background music LyricsMUTEMATH...   
82   Irrelevant Cabbage’s Ever-Growing CD/Vinyl Col...   
83   Lean On (NGHTMRE Remix) Lyrics[Verse 1: MØ]\nD...   
89   Lara LyricsLara[jb]\n\nCANTO THE FIRST[265]\ns...   
92   ​vwm’s immaculate collection LyricsVinylA Cert...   
93   Love Won’t Let Me Wait Lyrics[Verse 1]\nThe ti...   
96   Top Tracks of 2016 - Global LyricsDrake

#### Export them so my girl Giana can fix them (🥹 )

In [27]:
_2for_giana_to_correct = _non_lyrics_word_url_ending.merge(
    on='song_id', 
    right=songs_df[['song_id', 'song', 'performer', 'chart_debut']],
).drop(columns=['lyrics', 'song_id'])
_2for_giana_to_correct
save_lyrics(_2for_giana_to_correct, 'TEMP: for Giana to correct 2')

urls  \
0   [https://genius.com/Archie-bell-and-the-drells...   
1   [https://genius.com/Celedon-cq-soundtrack-of-m...   
2   [https://genius.com/Genius-june-2021-singles-r...   
3   [https://genius.com/Doja-cat-doja-cat-discogra...   
4   [https://genius.com/D-h-lawrence-lady-chatterl...   
5   [https://genius.com/Spotify-this-is-machine-gu...   
6   [https://genius.com/Formerly-vinylzombie-mcr-n...   
7   [https://genius.com/James-buchanan-state-of-th...   
8   [https://genius.com/Spotify-todays-top-hits-2-...   
9   [https://genius.com/Frank-mccourt-angelas-ashe...   
10  [https://genius.com/Spotify-todays-top-hits-2-...   
11  [https://genius.com/Oliver-stone-scarface-scri...   
12  [https://genius.com/Spotify-black-history-salu...   
13  [https://genius.com/Eleventh-cassette-2021-new...   
14  [https://genius.com/Grambling-state-football-l...   
15  [https://genius.com/Tom-jones-ill-never-fall-i...   
16  [https://genius.com/Vintagewashingmachine-vwms...   
17  [https://genius.com/Josiah-botting-youth-group...   
18  [https://genius.com/Irrelevant-cabbage-irrelev...   
19  [https://genius.com/Major-lazer-and-dj-snake-l...   
20  [https://genius.com/Lord-byron-lara-annotated,...   
21  [https://genius.com/Vintagewashingmachine-vwms...   
22  [https://genius.com/Major-harris-love-wont-let...   
23  [https://genius.com/Spotify-top-tracks-of-2016...   
24  [https://genius.com/Howard-zinn-chapter-11-rob...   
25  [https://genius.com/Jules-verne-in-search-of-t...   
26  [https://genius.com/Benjamin-franklin-the-auto...   
27  [https://genius.com/Montague-summers-the-vampi...   
28  [https://genius.com/Howard-zinn-chapter-21-car...   
29  [https://genius.com/Arthur-edward-waite-the-he...   
30  [https://genius.com/Grammys-2015-nominees-anno...   
31  [https://genius.com/Frank-mccourt-angelas-ashe...   
32  [https://genius.com/James-joyce-ulysses-chap-1...   
33  [https://genius.com/Frank-mccourt-angelas-ashe...   
34  [https://genius.com/James-joyce-ulysses-chap-1...   
35  [https://genius.com/Whatroughbeast-halloween-p...   
36  [https://genius.com/Onlyeverarabella-the-relat...   
37  [https://genius.com/Gerald-haywood-2018-haywoo...   
38  [https://genius.com/James-joyce-finnegans-wake...   
39  [https://genius.com/Bell-hooks-eating-the-othe...   
40  [https://genius.com/Mr-petrucci-how-to-write-a...   
41  [https://genius.com/Pitchfork-the-200-best-son...   
42  [https://genius.com/Sir-arthur-conan-doyle-cha...   
43  [https://genius.com/Rolling-stone-the-500-grea...   

                                                 song  \
0                                I Can't Stop Dancing   
1                                     I Gotta Feeling   
2                                  I Think I Love You   
3                                               34+35   
4                                 As If I Didn't Know   
5                                          Bad Things   
6                                       Boom Boom Pow   
7                                      Convention '72   
8                               Do We Have A Problem?   
9                  Doesn't Somebody Want To Be Wanted   
10                                              Enemy   
11                                 Every 1's A Winner   
12                                    Everyday People   
13         Gonna Make You Sweat (Everybody Dance Now)   
14                                         Here We Go   
15                      I'll Never Fall In Love Again   
16                                            Imagine   
17                                        It Ain't Me   
18                                              Layla   
19                                            Lean On   
20                                        Lost In You   
21                     Love Theme From Romeo & Juliet   
22                             Love Won't Let Me Wait   
23                                     Me, Myself & I   
24                             More, More, More Pt. 1   
2

# 5.  Manually (thanks, Giana!) fix missing / problematic URLs

### We found 179 likely wrong (cuz of url). POST RE-FETCH: 123 likely right now => 56 not covered. 44 "non `-lyrics`" urls, 12 failed

### ALSO: in the larger dataset there are 68 just missing URLS in general

### SO: (A) have Giana help with the 80 URL-less songs (+ 4 I manually found) (B) have her do the 44 "non `-lyrics`" URLS   

## A. Fix the 80 URL-less songs (existing + failed fetch)

### Ok, so these ones *could* be right... they just don't have a URL...

In [28]:
never_got_assigned_urls = lyrics_sans_instrumentals[
    lyrics_sans_instrumentals.urls.isna()
]
never_got_assigned_urls

song_id  \
4574  Drop It Like It's HotSnoop Dogg Featuring Phar...   
4575              Flashdance...What A FeelingIrene Cara   
4576            I Love ItIcona Pop Featuring Charli XCX   
4577      Let Me Blow Ya MindEve Featuring Gwen Stefani   
4578  Moves Like JaggerMaroon 5 Featuring Christina ...   
...                                                 ...   
4637           We Are Youngfun. Featuring Janelle Monae   
4638       We Found LoveRihanna Featuring Calvin Harris   
4639                                 We R Who We RKe$ha   
4640                    Young ForeverJay-Z + Mr. Hudson   
4641                          Your Love Is My DrugKe$ha   

                                                 lyrics urls  
4574  TranslationsPortuguêsDrop It Like It’s Hot Lyr...  NaN  
4575  Flashdance... What a Feeling Lyrics[Intro]\nFi...  NaN  
4576  TranslationsPortuguêsI Love It Lyrics[Chorus: ...  NaN  
4577  Let Me Blow Ya Mind Lyrics[Intro: Eve]\nUh, uh...  NaN  
4578  Moves Like Jagger Lyrics[Intro: Adam Levine]\n...  NaN  
...                                                 ...  ...  
4637  We Are Young Lyrics[Verse 1: Nate Ruess]\nGive...  NaN  
4638  TranslationsPortuguêsEnglishFrançaisWe Found L...  NaN  
4639  TranslationsPortuguêsWe R Who We R Lyrics[Vers...  NaN  
4640  Young Forever Lyrics[Intro: Mr. Hudson]\nLet's...  NaN  
4641  TranslationsPortuguêsYour Love is My Drug Lyri...  NaN  

[68 rows x 3 columns]

### These DEFINITELY need URLs cuz the fetch just failed on them

In [29]:
_refetched_df.urls.isna().sum()

12

In [30]:
_refetch_completely_failed = _refetched_df[
    _refetched_df.lyrics.isna()
]
_refetch_completely_failed.reset_index(drop=True)

song_id lyrics  urls
0   1-800-273-8255Logic Featuring Alessia Cara & K...    NaN  None
1               American Pie (Parts I & II)Don McLean    NaN  None
2        American Woman/No Sugar TonightThe Guess Who    NaN  None
3   As Long As You Love MeJustin Bieber Featuring ...    NaN  None
4               FancyIggy Azalea Featuring Charli XCX    NaN  None
5   It Must Have Been Love (From "Pretty Woman")Ro...    NaN  None
6                   Love Machine (Part 1)The Miracles    NaN  None
7                    Mood24kGoldn Featuring iann dior    NaN  None
8              One DanceDrake Featuring WizKid & Kyla    NaN  None
9       Say Say SayPaul McCartney And Michael Jackson    NaN  None
10  Take My Breath Away (Love Theme From "Top Gun"...    NaN  None
11    An Open Letter To My Teenage SonVictor Lundberg    NaN  None

### What I want Giana to correct part A

In [31]:
_for_giana_to_correct = pd.concat([
    never_got_assigned_urls,
    _refetch_completely_failed
]).drop_duplicates(subset=['song_id']).merge(
    on='song_id', 
    right=songs_df[['song_id', 'song', 'performer', 'chart_debut']],
).drop(columns=['lyrics', 'song_id'])
_for_giana_to_correct
# save_lyrics(_for_giana_to_correct, 'TEMP: for Giana to correct')

urls                                             song  \
0    NaN                            Drop It Like It's Hot   
1    NaN                      Flashdance...What A Feeling   
2    NaN                                        I Love It   
3    NaN                              Let Me Blow Ya Mind   
4    NaN                                Moves Like Jagger   
..   ...                                              ...   
75  None                                             Mood   
76  None                                        One Dance   
77  None                                      Say Say Say   
78  None  Take My Breath Away (Love Theme From "Top Gun")   
79  None                 An Open Letter To My Teenage Son   

                                performer chart_debut  
0           Snoop Dogg Featuring Pharrell  2004-10-02  
1                              Irene Cara  1983-04-02  
2          Icona Pop Featuring Charli XCX  2013-02-16  
3              Eve Featuring Gwen Stefani  2001-04-28  
4   Maroon 5 Featuring Christina Aguilera  2011-07-09  
..                                    ...         ...  
75           24kGoldn Featuring iann dior  2020-08-22  
76          Drake Featuring WizKid & Kyla  2016-04-23  
77     Paul McCartney And Michael Jackson  1983-10-15  
78                                 Berlin  1986-06-21  
79                        Victor Lundberg  1967-11-11  

[80 rows x 4 columns]

### ✅ (A) NEW URLS ADDED BY GIANA. Now download the lyrics & merge them!

Note: I found 80 problematic URLs progamatically, and added 7 found via viz inspection

### note... WE'RE DROPPING some gems: 
# NEW NOTE: actually I'm going to add these back in a later notebook lol
- "Convention '72" is this hilarious mashup of a political inteview & random / irreverent snippets of songs lol https://www.youtube.com/watch?v=gofKPwPDAg8

In [32]:
urls_fixed_by_giana_RAW = pd.read_csv(f'{LYRIC_FOLDER_PATH}/{"urls corrected by Giana Ded 14 2022"}.csv')

urls_fixed_by_giana_RAW['song_id'] = urls_fixed_by_giana_RAW.song + urls_fixed_by_giana_RAW.performer
urls_fixed_by_giana_RAW = urls_fixed_by_giana_RAW.drop(
        columns=['song', 'performer', 'chart_debut']
    ).rename(
        columns={'PASTE LYRICS URL HERE': 'urls'}
    )

urls_fixed_by_giana = urls_fixed_by_giana_RAW[
    urls_fixed_by_giana_RAW.urls != IS_INSTRUMENTAL_RETURN_FLAG
]
urls_fixed_by_giana

urls  \
0   https://genius.com/Snoop-dogg-drop-it-like-its...   
1   https://genius.com/Irene-cara-flashdance-what-...   
2       https://genius.com/Icona-pop-i-love-it-lyrics   
3   https://genius.com/Eve-let-me-blow-ya-mind-lyrics   
4   https://genius.com/Maroon-5-moves-like-jagger-...   
..                                                ...   
82  https://genius.com/The-four-seasons-ive-got-yo...   
83  https://genius.com/Linda-ronstadt-and-james-in...   
84            https://genius.com/Dram-broccoli-lyrics   
85        https://genius.com/Fabolous-into-you-lyrics   
86        https://genius.com/Bob-nothin-on-you-lyrics   

                                              song_id  
0   Drop It Like It's HotSnoop Dogg Featuring Phar...  
1               Flashdance...What A FeelingIrene Cara  
2             I Love ItIcona Pop Featuring Charli XCX  
3       Let Me Blow Ya MindEve Featuring Gwen Stefani  
4   Moves Like JaggerMaroon 5 Featuring Christina ...  
..                                                ...  
82  The 4 Seasons Featuring the "Sound of Frankie ...  
83  Somewhere Out There (From "An American Tail")L...  
84              BroccoliD.R.A.M. Featuring Lil Yachty  
85        Into YouFabolous Featuring Tamia Or Ashanti  
86           Nothin' On YouB.o.B Featuring Bruno Mars  

[84 rows x 2 columns]

In [33]:
def _fetch_lyrics_for_song(urls):
    try: 
        print(urls[0])
        return [genius.lyrics(song_url=urls[0]), urls]
    except:
        print(f'OOPS!')
        
        
def _fetch_lyrics_via_url(songs_subset_df):
    return songs_subset_df.apply(
        lambda row: _fetch_lyrics_for_song(urls=[row['urls']]),
        axis=1
    )

In [34]:
# _4giana_refetch = _fetch_lyrics_via_url(urls_fixed_by_giana)

In [35]:
# fetched_giana_urls = convert_fetched_lyrics_to_df(_4giana_refetch, urls_fixed_by_giana)
# fetched_giana_urls

In [36]:
# save_lyrics(fetched_giana_urls, 'URLs from Giana, with the lyrics fetched for them')
fetched_giana_urls1 = pd.read_csv(f'{LYRIC_FOLDER_PATH}/{"URLs from Giana, with the lyrics fetched for them"}.csv')
fetched_giana_urls1

song_id  \
0   Drop It Like It's HotSnoop Dogg Featuring Phar...   
1               Flashdance...What A FeelingIrene Cara   
2             I Love ItIcona Pop Featuring Charli XCX   
3       Let Me Blow Ya MindEve Featuring Gwen Stefani   
4   Moves Like JaggerMaroon 5 Featuring Christina ...   
..                                                ...   
79  The 4 Seasons Featuring the "Sound of Frankie ...   
80  Somewhere Out There (From "An American Tail")L...   
81              BroccoliD.R.A.M. Featuring Lil Yachty   
82        Into YouFabolous Featuring Tamia Or Ashanti   
83           Nothin' On YouB.o.B Featuring Bruno Mars   

                                               lyrics  \
0   TranslationsPortuguêsDrop It Like It’s Hot Lyr...   
1   Flashdance... What a Feeling Lyrics[Intro]\nFi...   
2   TranslationsPortuguêsI Love It Lyrics[Chorus: ...   
3   Let Me Blow Ya Mind Lyrics[Intro: Eve]\nUh, uh...   
4   Moves Like Jagger Lyrics[Intro: Adam Levine]\n...   
..                                                ...   
79  I’ve Got You Under My Skin LyricsI've got you ...   
80  Somewhere Out There Lyrics[Verse 1: Linda Rons...   
81  Broccoli Lyrics[Intro: D.R.A.M.]\nAin't no tel...   
82  Into You Lyrics[Intro: Fabolous & Ashanti]\nBa...   
83  Nothin’ on You Lyrics[Intro: Bruno Mars & B.o....   

                                                 urls  
0   ["https://genius.com/Snoop-dogg-drop-it-like-i...  
1   ["https://genius.com/Irene-cara-flashdance-wha...  
2   ["https://genius.com/Icona-pop-i-love-it-lyrics"]  
3   ["https://genius.com/Eve-let-me-blow-ya-mind-l...  
4   ["https://genius.com/Maroon-5-moves-like-jagge...  
..                                                ...  
79  ["https://genius.com/The-four-seasons-ive-got-...  
80  ["https://genius.com/Linda-ronstadt-and-james-...  
81        ["https://genius.com/Dram-broccoli-lyrics"]  
82    ["https://genius.com/Fabolous-into-you-lyrics"]  
83    ["https://genius.com/Bob-nothin-on-you-lyrics"]  

[84 rows x 3 columns]

## PART (B): refetch lyrics (using Giana's correct URLs 2) for the 44 songs whose URL did *not* previously end in `-lyrics`

In [37]:
urls2_fixed_by_giana_RAW = pd.read_csv(f'{LYRIC_FOLDER_PATH}/{"urls corrected by Giana 2 Dec 15 2022"}.csv')

urls2_fixed_by_giana_RAW['song_id'] = urls2_fixed_by_giana_RAW.song + urls2_fixed_by_giana_RAW.performer
urls2_fixed_by_giana_RAW = urls2_fixed_by_giana_RAW.drop(
        columns=['song', 'performer', 'chart_debut']
    ).rename(
        columns={'ADD URL HERE': 'urls'}
    )

urls2_fixed_by_giana = urls2_fixed_by_giana_RAW[urls2_fixed_by_giana_RAW.urls != IS_INSTRUMENTAL_RETURN_FLAG]
urls2_fixed_by_giana

urls  \
0   https://genius.com/Archie-bell-i-cant-stop-dan...   
1   https://genius.com/Black-eyed-peas-i-gotta-fee...   
2   https://genius.com/The-partridge-family-i-thin...   
3       https://genius.com/Ariana-grande-34-35-lyrics   
4   https://genius.com/Adam-wade-as-if-i-didnt-kno...   
5   https://genius.com/Machine-gun-kelly-and-camil...   
6   https://genius.com/Black-eyed-peas-boom-boom-p...   
8   https://genius.com/Nicki-minaj-and-lil-baby-do...   
9   https://genius.com/The-partridge-family-doesnt...   
10  https://genius.com/Imagine-dragons-and-jid-ene...   
11  https://genius.com/Hot-chocolate-every-1s-a-wi...   
12  https://genius.com/Sly-and-the-family-stone-ev...   
13  https://genius.com/C-c-music-factory-gonna-mak...   
14  https://genius.com/C-c-music-factory-here-we-g...   
15  https://genius.com/Tom-jones-ill-never-fall-in...   
16      https://genius.com/John-lennon-imagine-lyrics   
17  https://genius.com/Kygo-and-selena-gomez-it-ai...   
18  https://genius.com/Derek-and-the-dominos-layla...   
19      https://genius.com/Major-lazer-lean-on-lyrics   
20  https://genius.com/Garth-brooks-lost-in-you-ly...   
21  https://genius.com/Nino-rota-what-is-a-youth-l...   
22  https://genius.com/Major-harris-love-wont-let-...   
23  https://genius.com/G-eazy-and-bebe-rexha-me-my...   
24  https://genius.com/Andrea-true-connection-more...   
26  https://genius.com/Elton-john-philadelphia-fre...   
28  https://genius.com/Esther-phillips-release-me-...   
29   https://genius.com/Enigma-sadeness-part-i-lyrics   
30  https://genius.com/A-great-big-world-say-somet...   
31  https://genius.com/Bob-seger-shame-on-the-moon...   
32  https://genius.com/1910-fruitgum-company-simon...   
33  https://genius.com/Bob-seger-still-the-same-ly...   
35  https://genius.com/Ben-e-king-supernatural-thi...   
36  https://genius.com/Barry-sadler-ballad-of-the-...   
37  https://genius.com/Toni-fisher-the-big-hurt-ly...   
39  https://genius.com/Joey-scarbury-believe-it-or...   
40  https://genius.com/Black-eyed-peas-where-is-th...   
41  https://genius.com/The-shirelles-will-you-stil...   
42  https://genius.com/Arthur-lyman-yellow-bird-ly...   
43  https://genius.com/Five-stairsteps-o-o-h-child...   

                                              song_id  
0        I Can't Stop DancingArchie Bell & The Drells  
1                  I Gotta FeelingThe Black Eyed Peas  
2   I Think I Love YouThe Partridge Family Starrin...  
3   34+35Ariana Grande Feat. Doja Cat & Megan Thee...  
4                        As If I Didn't KnowAdam Wade  
5        Bad ThingsMachine Gun Kelly x Camila Cabello  
6                    Boom Boom PowThe Black Eyed Peas  
8         Do We Have A Problem?Nicki Minaj X Lil Baby  
9   Doesn't Somebody Want To Be WantedThe Partridg...  
10                         EnemyImagine Dragons X JID  
11                    Every 1's A WinnerHot Chocolate  
12              Everyday PeopleSly & The Family Stone  
13  Gonna Make You Sweat (Everybody Dance Now)C+C ...  
14            Here We GoC+C Music Factory/F. Williams  
15             I'll Never Fall In Love AgainTom Jones  
16                ImagineJohn Lennon Plastic Ono Band  
17                     It Ain't MeKygo x Selena Gomez  
18                           LaylaDerek & The Dominos  
19         Lean OnMajor Lazer & DJ Snake Featuring MO  
20            Lost In YouGarth Brooks as Chris Gaines  
21  Love Theme From Romeo & JulietHenry Mancini An...  
22                 Love Won't Let Me WaitMajor Harris  
23                  Me, Myself & IG-Eazy x Bebe Rexha  
24       More, More, More Pt. 1Andrea True Connection  
26            Philadelphia FreedomThe Elton John Band  
28          Release MeEsther Phillips "Little Esther"  
29                              Sadeness Part 1Enigma  
30  Say SomethingA Great Big World & Christina Agu...  
31  Shame On The MoonBob Seger & The Silver Bullet...  
32                        Simon Says1910 Fruitgum Co.  
33   Still The SameBob Seger & T

In [38]:
# _giana2_refetch = _fetch_lyrics_via_url(urls2_fixed_by_giana)

In [39]:
# fetched2_giana_urls = convert_fetched_lyrics_to_df(_giana2_refetch, urls2_fixed_by_giana)
# fetched2_giana_urls

In [40]:
# save_lyrics(fetched2_giana_urls, 'URLs from Giana 2, with the lyrics fetched for them')
fetched_giana_urls2 = pd.read_csv(f'{LYRIC_FOLDER_PATH}/{"URLs from Giana 2, with the lyrics fetched for them"}.csv')
fetched_giana_urls2

song_id  \
0        I Can't Stop DancingArchie Bell & The Drells   
1                  I Gotta FeelingThe Black Eyed Peas   
2   I Think I Love YouThe Partridge Family Starrin...   
3   34+35Ariana Grande Feat. Doja Cat & Megan Thee...   
4                        As If I Didn't KnowAdam Wade   
5        Bad ThingsMachine Gun Kelly x Camila Cabello   
6                    Boom Boom PowThe Black Eyed Peas   
7         Do We Have A Problem?Nicki Minaj X Lil Baby   
8   Doesn't Somebody Want To Be WantedThe Partridg...   
9                          EnemyImagine Dragons X JID   
10                    Every 1's A WinnerHot Chocolate   
11              Everyday PeopleSly & The Family Stone   
12  Gonna Make You Sweat (Everybody Dance Now)C+C ...   
13            Here We GoC+C Music Factory/F. Williams   
14             I'll Never Fall In Love AgainTom Jones   
15                ImagineJohn Lennon Plastic Ono Band   
16                     It Ain't MeKygo x Selena Gomez   
17                           LaylaDerek & The Dominos   
18         Lean OnMajor Lazer & DJ Snake Featuring MO   
19            Lost In YouGarth Brooks as Chris Gaines   
20  Love Theme From Romeo & JulietHenry Mancini An...   
21                 Love Won't Let Me WaitMajor Harris   
22                  Me, Myself & IG-Eazy x Bebe Rexha   
23       More, More, More Pt. 1Andrea True Connection   
24            Philadelphia FreedomThe Elton John Band   
25          Release MeEsther Phillips "Little Esther"   
26                              Sadeness Part 1Enigma   
27  Say SomethingA Great Big World & Christina Agu...   
28  Shame On The MoonBob Seger & The Silver Bullet...   
29                        Simon Says1910 Fruitgum Co.   
30   Still The SameBob Seger & The Silver Bullet Band   
31             Supernatural Thing - Part IBen E. King   
32    The Ballad Of The Green BeretsSSgt Barry Sadler   
33                       The Big HurtMiss Toni Fisher   
34  Theme From "Greatest American Hero" (Believe I...   
35              Where Is The Love?The Black Eyed Peas   
36             Will You Love Me TomorrowThe Shirelles   
37                      Yellow BirdArthur Lyman Group   
38          O-o-h Child/Dear PrudenceThe 5 Stairsteps   

                                               lyrics  \
0   I Can’t Stop Dancing Lyrics(Oh, no, I just can...   
1   TranslationsEnglishI Gotta Feeling Lyrics[Chor...   
2   I Think I Love You Lyrics[Intro]\nBaaa, ba ba ...   
3   TranslationsPolskiTürkçeEspañolHrvatskiPortugu...   
4   As If I Didn’t Know Lyrics[Verse 1]\nYou swore...   
5   Bad Things Lyrics[Chorus: Camila Cabello]\nAm ...   
6   Boom Boom Pow Lyrics[Album Intro]\nWelcome\nWe...   
7   TranslationsTürkçeEspañolРусскийPortuguêsDo We...   
8   Doesn’t Somebody Want To Be Wanted Lyrics[Intr...   
9   TranslationsEspañolEnglishDeutschTagalog / ᜆᜄᜎ...   
10  Every 1's a Winner LyricsNever could believe t...   
11  Everyday People Lyrics[Verse 1]\nSometimes I'm...   
12  Gonna Make You Sweat (Everybody Dance Now) Lyr...   
13  Here We Go, Let’s Rock and Roll LyricsYou all ...   
14  I’ll Never Fall in Love Again Lyrics[Vers 1]\n...   
15  Imagine Lyrics[Verse 1]\nImagine there's no he...   
16  TranslationsDeutschFrançaisIt Ain’t Me Lyrics[...   
17  Layla Lyrics[Opening Instrumental Riff]\n\n[Ve...   
18  TranslationsPortuguêsFrançaisΕλληνικάLean On L...   
19  Lost In You Lyrics[Verse]\nThere's no more wai...   
20  What Is a Youth (Love Theme From ’Romeo & Juli...   
21  Love Won’t Let Me Wait Lyrics[Verse 1]\nThe ti...   
22  TranslationsFrançaisMe, Myself & I Lyrics[Chor...   
23  More, More, More Lyrics[Intro]\nOoh, how do yo...   
24  Philadelphia Freedom Lyrics[Verse 1]\nI used t...   
25  Release Me LyricsLet me go, let me go, let me ...   
26  Sadeness (Part I) LyricsProcedamus in pace\nIn...   
27  TranslationsPortuguêsSay Something Lyrics[Intr...   
28  Shame on the Moon Lyrics[Verse 1]\nUntil you'v...   
29  Simon Says Lyrics{Instrumental}\nI'd like to p...  

# 7. Merge Giana-assisted fixes into the main dataset! (and update the list of instrumentals)

## First, identify the instrumentals from this notebook, update total record

In [41]:
NEW_INSTRUMENTALS_FOUND

['Our Winter LoveBill Pursell']

In [42]:
instrumentals_from_giana = urls_fixed_by_giana_RAW[urls_fixed_by_giana_RAW.urls == IS_INSTRUMENTAL_RETURN_FLAG].song_id.values


In [43]:
instrumentals2_from_giana = urls2_fixed_by_giana_RAW[urls2_fixed_by_giana_RAW.urls == IS_INSTRUMENTAL_RETURN_FLAG].song_id.values


In [44]:
ALL_NEW_INSTRUMENTALS = [*NEW_INSTRUMENTALS_FOUND, *list(instrumentals_from_giana), *list(instrumentals2_from_giana)]
ALL_NEW_INSTRUMENTALS

['Our Winter LoveBill Pursell',
 'Near YouRoger Williams',
 'An Open Letter To My Teenage SonVictor Lundberg',
 'Peter GunnRay Anthony and His Orchestra',
 "Convention '72The Delegates",
 'MoreKai Winding & Orchestra',
 'Quiet VillageThe Exotic Sounds of Martin Denny',
 'Stranger On The ShoreMr. Acker Bilk',
 'The Theme From Hill Street BluesMike Post featuring Larry Carlton']

In [45]:
new_instrumentals_df = pd.DataFrame({'song_id': ALL_NEW_INSTRUMENTALS, 'lyrics': IS_INSTRUMENTAL_RETURN_FLAG, 'urls': None})
new_instrumentals_df


song_id                   lyrics  \
0                        Our Winter LoveBill Pursell  THIS_IS_AN_INSTRUMENTAL   
1                             Near YouRoger Williams  THIS_IS_AN_INSTRUMENTAL   
2    An Open Letter To My Teenage SonVictor Lundberg  THIS_IS_AN_INSTRUMENTAL   
3            Peter GunnRay Anthony and His Orchestra  THIS_IS_AN_INSTRUMENTAL   
4                        Convention '72The Delegates  THIS_IS_AN_INSTRUMENTAL   
5                        MoreKai Winding & Orchestra  THIS_IS_AN_INSTRUMENTAL   
6     Quiet VillageThe Exotic Sounds of Martin Denny  THIS_IS_AN_INSTRUMENTAL   
7                Stranger On The ShoreMr. Acker Bilk  THIS_IS_AN_INSTRUMENTAL   
8  The Theme From Hill Street BluesMike Post feat...  THIS_IS_AN_INSTRUMENTAL   

   urls  
0  None  
1  None  
2  None  
3  None  
4  None  
5  None  
6  None  
7  None  
8  None

In [46]:
INSTR_FILE = "INSTRUMENTAL SONGS final"
_existing_istrumentals = pd.read_csv(f'{LYRIC_FOLDER_PATH}/{INSTR_FILE}.csv')
updated_instr = pd.concat([_existing_istrumentals, new_instrumentals_df]).drop_duplicates()
updated_instr
save_lyrics(updated_instr, INSTR_FILE)

song_id  \
0                                     95.southJ. Cole   
1   A Fifth Of BeethovenWalter Murphy & The Big Ap...   
2                            Axel FHarold Faltermeyer   
3   Can't Hold UsMacklemore & Ryan Lewis Featuring...   
4                         Classical GasMason Williams   
..                                                ...   
64     Quiet VillageThe Exotic Sounds of Martin Denny   
65                Stranger On The ShoreMr. Acker Bilk   
66  The Theme From Hill Street BluesMike Post feat...   
67       Percolator (Twist)Billy Joe & The Checkmates   
68    An Open Letter To My Teenage SonVictor Lundberg   

                     lyrics                                               urls  
0   THIS_IS_AN_INSTRUMENTAL  ["https://genius.com/8-bit-arcade-95-south-8-b...  
1   THIS_IS_AN_INSTRUMENTAL  ["https://genius.com/Walter-murphy-a-fifth-of-...  
2   THIS_IS_AN_INSTRUMENTAL  ["https://genius.com/Harold-faltermeyer-axel-f...  
3   THIS_IS_AN_INSTRUMENTAL  ["https://genius.com/Cant-hold-us-cant-hold-us...  
4   THIS_IS_AN_INSTRUMENTAL  ["https://genius.com/Mason-williams-classical-...  
..                      ...                                                ...  
64  THIS_IS_AN_INSTRUMENTAL                                                NaN  
65  THIS_IS_AN_INSTRUMENTAL                                                NaN  
66  THIS_IS_AN_INSTRUMENTAL                                                NaN  
67  THIS_IS_AN_INSTRUMENTAL                                                NaN  
68  THIS_IS_AN_INSTRUMENTAL                                                NaN  

[69 rows x 3 columns]

## Second, merge the fixed data in

In [47]:
giana_updates_all = pd.concat([fetched_giana_urls1, fetched_giana_urls2]).drop_duplicates()
giana_updates_all

song_id  \
0   Drop It Like It's HotSnoop Dogg Featuring Phar...   
1               Flashdance...What A FeelingIrene Cara   
2             I Love ItIcona Pop Featuring Charli XCX   
3       Let Me Blow Ya MindEve Featuring Gwen Stefani   
4   Moves Like JaggerMaroon 5 Featuring Christina ...   
..                                                ...   
34  Theme From "Greatest American Hero" (Believe I...   
35              Where Is The Love?The Black Eyed Peas   
36             Will You Love Me TomorrowThe Shirelles   
37                      Yellow BirdArthur Lyman Group   
38          O-o-h Child/Dear PrudenceThe 5 Stairsteps   

                                               lyrics  \
0   TranslationsPortuguêsDrop It Like It’s Hot Lyr...   
1   Flashdance... What a Feeling Lyrics[Intro]\nFi...   
2   TranslationsPortuguêsI Love It Lyrics[Chorus: ...   
3   Let Me Blow Ya Mind Lyrics[Intro: Eve]\nUh, uh...   
4   Moves Like Jagger Lyrics[Intro: Adam Levine]\n...   
..                                                ...   
34  Believe It or Not LyricsLook at what's happene...   
35  TranslationsPortuguêsEnglishWhere Is the Love?...   
36  Will You Still Love Me Tomorrow Lyrics[Verse 1...   
37  Yellow Bird LyricsYellow bird, yellow bird\nI'...   
38  O-o-h Child Lyrics[Verse 1]\nOoh-oo child\nThi...   

                                                 urls  
0   ["https://genius.com/Snoop-dogg-drop-it-like-i...  
1   ["https://genius.com/Irene-cara-flashdance-wha...  
2   ["https://genius.com/Icona-pop-i-love-it-lyrics"]  
3   ["https://genius.com/Eve-let-me-blow-ya-mind-l...  
4   ["https://genius.com/Maroon-5-moves-like-jagge...  
..                                                ...  
34  ["https://genius.com/Joey-scarbury-believe-it-...  
35  ["https://genius.com/Black-eyed-peas-where-is-...  
36  ["https://genius.com/The-shirelles-will-you-st...  
37  ["https://genius.com/Arthur-lyman-yellow-bird-...  
38  ["https://genius.com/Five-stairsteps-o-o-h-chi...  

[123 rows x 3 columns]

In [48]:
_merged_giana = pd.concat([
    remove_overlap(ready_for_manual_fixes, giana_updates_all),
    giana_updates_all
]).reset_index(drop=True)

ready_for_export = _merged_giana[
    _merged_giana.song_id.apply(lambda song_id: song_id not in updated_instr.song_id.values)
].drop_duplicates(subset='song_id') # to prove it to ourselves

ready_for_export['urls'] = ready_for_export['urls'].apply(
    lambda urls: urls if isinstance(urls, list) else json.loads(urls)
)
ready_for_export

song_id  \
0       (I Can't Get No) SatisfactionThe Rolling Stones   
1                (I Just) Died In Your ArmsCutting Crew   
2                (I Know) I'm Losing YouThe Temptations   
3     (I've Had) The Time Of My LifeBill Medley & Je...   
4                                         1 ThingAmerie   
...                                                 ...   
4996  Theme From "Greatest American Hero" (Believe I...   
4997              Where Is The Love?The Black Eyed Peas   
4998             Will You Love Me TomorrowThe Shirelles   
4999                      Yellow BirdArthur Lyman Group   
5000          O-o-h Child/Dear PrudenceThe 5 Stairsteps   

                                                 lyrics  \
0     (I Can’t Get No) Satisfaction Lyrics[Instrumen...   
1     (I Just) Died in Your Arms Lyrics[Intro]\nOh I...   
2     (I Know) I’m Losing You Lyrics[Intro]\nOoh\nYo...   
3     (I’ve Had) The Time of My Life Lyrics[Chorus: ...   
4     1 Thing Lyrics[Intro]\nWoo!\nUh\nWoo!\nNa-na-n...   
...                                                 ...   
4996  Believe It or Not LyricsLook at what's happene...   
4997  TranslationsPortuguêsEnglishWhere Is the Love?...   
4998  Will You Still Love Me Tomorrow Lyrics[Verse 1...   
4999  Yellow Bird LyricsYellow bird, yellow bird\nI'...   
5000  O-o-h Child Lyrics[Verse 1]\nOoh-oo child\nThi...   

                                                   urls  
0     [https://genius.com/The-rolling-stones-i-cant-...  
1     [https://genius.com/Cutting-crew-i-just-died-i...  
2     [https://genius.com/The-temptations-i-know-im-...  
3     [https://genius.com/Bill-medley-and-jennifer-w...  
4     [https://genius.com/Amerie-1-thing-lyrics, htt...  
...                                                 ...  
4996  [https://genius.com/Joey-scarbury-believe-it-o...  
4997  [https://genius.com/Black-eyed-peas-where-is-t...  
4998  [https://genius.com/The-shirelles-will-you-sti...  
4999  [https://genius.com/Arthur-lyman-yellow-bird-l...  
5000  [https://genius.com/Five-stairsteps-o-o-h-chil...  

[4998 rows x 3 columns]

# Sanity checks:

In [49]:
#sanity check
ready_for_export.lyrics.isna().sum() + ready_for_export.urls.isna().sum()

0

### Manual inspection of these non `-lyrics` URLS shows they are correct 😊  

In [50]:
ready_for_export[
    ready_for_export.urls.apply(lambda url: not url[0].endswith('-lyrics'))
].urls.apply(lambda urls: urls[0]).values

array(['https://genius.com/Tom-jones-ill-never-fall-in-love-again-annotated',
       'https://genius.com/Major-harris-love-wont-let-me-wait-annotated'],
      dtype=object)

# Export

In [51]:
LYRICS_DATASET = '2.4 OUTPUT lyrics for all songs sans instrumentals'
save_lyrics(ready_for_export, LYRICS_DATASET)

song_id  \
0       (I Can't Get No) SatisfactionThe Rolling Stones   
1                (I Just) Died In Your ArmsCutting Crew   
2                (I Know) I'm Losing YouThe Temptations   
3     (I've Had) The Time Of My LifeBill Medley & Je...   
4                                         1 ThingAmerie   
...                                                 ...   
4996  Theme From "Greatest American Hero" (Believe I...   
4997              Where Is The Love?The Black Eyed Peas   
4998             Will You Love Me TomorrowThe Shirelles   
4999                      Yellow BirdArthur Lyman Group   
5000          O-o-h Child/Dear PrudenceThe 5 Stairsteps   

                                                 lyrics  \
0     (I Can’t Get No) Satisfaction Lyrics[Instrumen...   
1     (I Just) Died in Your Arms Lyrics[Intro]\nOh I...   
2     (I Know) I’m Losing You Lyrics[Intro]\nOoh\nYo...   
3     (I’ve Had) The Time of My Life Lyrics[Chorus: ...   
4     1 Thing Lyrics[Intro]\nWoo!\nUh\nWoo!\nNa-na-n...   
...                                                 ...   
4996  Believe It or Not LyricsLook at what's happene...   
4997  TranslationsPortuguêsEnglishWhere Is the Love?...   
4998  Will You Still Love Me Tomorrow Lyrics[Verse 1...   
4999  Yellow Bird LyricsYellow bird, yellow bird\nI'...   
5000  O-o-h Child Lyrics[Verse 1]\nOoh-oo child\nThi...   

                                                   urls  
0     [https://genius.com/The-rolling-stones-i-cant-...  
1     [https://genius.com/Cutting-crew-i-just-died-i...  
2     [https://genius.com/The-temptations-i-know-im-...  
3     [https://genius.com/Bill-medley-and-jennifer-w...  
4     [https://genius.com/Amerie-1-thing-lyrics, htt...  
...                                                 ...  
4996  [https://genius.com/Joey-scarbury-believe-it-o...  
4997  [https://genius.com/Black-eyed-peas-where-is-t...  
4998  [https://genius.com/The-shirelles-will-you-sti...  
4999  [https://genius.com/Arthur-lyman-yellow-bird-l...  
5000  [https://genius.com/Five-stairsteps-o-o-h-chil...  

[4998 rows x 3 columns]